src/gleam/list.gleam

```diff
@@ -248,6 +250,67 @@ pub fn rest(list: List(a)) -> Result(List(a), Nil) {
   }
 }
 
+fn insert_to_group(
+  groups: Map(k, List(v)),
+  new_key: k,
+  new_value: v,
+) -> Map(k, List(v)) {
+  map.update(
+    groups,
+    new_key,
+    fn(v) {
+      case v {
+        Some(value) -> [new_value, ..value]
+        None -> [new_value]
+      }
+    },
+  )
+}
+
+/// Takes a list and groups the values by a key
+/// which is built from a key_selector function.
+///
+/// ## Examples
+///
+/// ```gleam
+/// > [Ok(3), Error("Wrong"), Ok(200), Ok(73)]
+///   |> group(with: fn(i) {
+///     case i {
+///       Ok(_) -> "Successful"
+///       Error(_) -> "Failed"
+///     }
+///   })
+///   |> map.to_list
+///
+/// [
+///   #("Failed", [Error("Wrong")]),
+///   #("Successful", [Ok(3), Ok(200), Ok(73)])
+/// ]
+///
+/// > group(from: [1,2,3,4,5], with: fn(i) {fn(i) { i - i / 3 * 3 }})
+/// |> map.to_list
+/// [#(0, [3]), #(1, [1, 4]), #(2, [2, 5])]
+/// ```
+///
+pub fn group(
+  from list: List(v),
+  with key_selector: fn(v) -> k,
+) -> Map(k, List(v)) {
+  map(list, fn(x) { #(key_selector(x), x) })
+  |> map_fold_right(map.new(), insert_to_group)
```

Comment:
Let's do this in one pass rather than two, and let's call this `do_group` to match the other helper function. I think it clearer than the current name, it's not a generic fold right over a map.